### ISSUE TITLE
Document the need for additional \-escaping of double quotes (") when calling external programs

### ISSUE THREAD

**Issue by mklement0** [MicrosoftDocs/PowerShell-Docs#2361]
Related: #5152; once https://github.com/PowerShell/PowerShell/issues/1995 gets fixed, this topic will become obsolete.

An _additional_ layer of escaping of `"` chars. is unexpectedly needed when calling external programs from PowerShell to make sure that the external program sees `"` embedded in `"..."` as `\"`, which is what PowerShell's own CLI expects too.

This is currently not documented, as far I can tell, neither in [about_Parsing](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_parsing) nor in [about_Quoting_Rules](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_quoting_rules) nor in [about_Special_Characters](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_special_characters)

_Update_: As an alternative to the manual `\`-escaping detailed below, you can use the **PSv3+ `ie` helper function** from the  **[`Native` module](https://github.com/mklement0/Native)** (in PSv5+, install with `Install-Module Native` from the [PowerShell Gallery](https://www.powershellgallery.com/packages/Native)), which internally **compensates for all broken behavior** and allows passing arguments as expected; to use it, simply prepend `ie` to your invocations; e.g.:  
`ie pwsh -noprofile -command ' "hi" '` works as expected.

---

**Calling an external program _from PowerShell_** (the external program just happens to be another instance of PowerShell (`pwsh`) in the following examples):

In order to pass string literal `"hi"` as a command to an external program _from PowerShell_, **PowerShell's own escaping of embedded `"` chars. must unexpectedly be supplemented with `\`-escaping** (the following commands are equivalent):

```powershell
# !! The \-escaping of the " chars. should NOT be necessary; without it, these commands would fail.
PS> pwsh -noprofile -command " \""hi\"" " 
PS> pwsh -noprofile -command " \`"hi\`" "
PS> pwsh -noprofile -command ' \"hi\" '  
hi
```

<sup>As an aside: If you're really calling another PowerShell instance from inside PowerShell, using a _script block_ (`{ ... }`) to pass the command avoids all quoting headaches and additionally enables support for _typed_ results (not just strings), albeit with the same limitations on type fidelity as with background jobs / remoting.</sup>

Note: There are also a number of **bugs**:

* Empty-string arguments are quietly _removed_ from the invocation.

* Values with embedded `"` are passed incorrectly: in addition to not automatically escaping them (the problem shown above), the situationally necessary enclosure in `"..."` behind the scenes is not reliably triggered, such as with
`3" of snow`.

* In Windows PowerShell only, values with spaces that end in a `\` char. are passed incorrectly.

For details and workarounds, see https://github.com/PowerShell/PowerShell/issues/1995#issuecomment-562334606

---

**Calling the PowerShell CLI _from another shell_** - `cmd.exe` or `bash`:

Note: The difficulties discussed next stem primarily from `cmd.exe` limitations and how commands are invoked by a single command-line string on Window. 
Calling from Bash / POSIX-like shells works robustly and as expected.

---

`cmd.exe`:

On Windows, PowerShell _Core_ now properly recognizes as `""` as escaped `"`, which enables robust  escaping, given that `""` is also recognized by `cmd.exe` itself:

```sh
# PowerShell *Core* only, on Windows only: use "", which works robustly.
C:\> pwsh.exe -noprofile -command " ""hi   &   dry"" " 
hi  &  dry  # OK
```

Sadly, `""` doesn't work in _Windows PowerShell_, where `\`-escaping for PowerShell's sake is _also_ required, which causes problems:

Using `\""` (sic) doesn't require escaping of cmd.exe metachars., but doesn't preserve whitespace as-is:

```sh
# No extra escaping needed, but whitespace is normalized.
C:\> powershell.exe -noprofile -command " \""hi   &   dry\"" " 
hi & dry   # !! Two spaces were collapsed into one.
```

Only using `\"` preserves whitespace as-is, but it additionally _requires individual `^`-escaping of the following `cmd.exe` metacharacters_ inside `\"...\"` runs: `& | < > ^`

```sh
# Whitespace is faithfully preserved, but cmd.exe metachars. must be ^-escaped
C:\> powershell.exe -noprofile -command " \"hi   ^&   dry\" " 
hi  &  dry  # OK
```

---

`bash`:

Use `\"` inside `"...."` strings, and no escaping at all inside `'...'` strings:

```sh
$ pwsh -noprofile -command " \"hi\" " # \-escaping needed for Bash's own sake (nesting ")
$ pwsh -noprofile -command ' "hi" ' # !! NO escaping of " needed
```

Note that on Unix-like platforms PowerShell's own command-line parsing does _not_ come into play (arguments are invariably passed as an _array_ of _literal_ tokens), and the above commands _solely_ use Bash [non]-escaping to pass an argument with literal contents ` "hi" `.

Thus:

* `\"` is Bash's native way to escape `"` inside a `"..."` string.
* Inside `'...'`, `"` needs no escaping
  * Caveat: if you tried `\"` inside `'...'`, the `\` chars. would be passed as _literals_, and the PowerShell command would break.

This asymmetry with how things work on Windows is unfortunate, but unavoidable.

Note that calling _from PowerShell_ on Unix-like platforms _still_ requires the extra `\`-escaping:

```powershell
PS> bash -c 'echo "one two" '  # !! only prints 'one', because the " are *ignored*
PS> bash -c 'echo \"one two\" '  # OK, but the \-escaping shouldn't be necessary.
```

While this requirement at least makes for consistent behavior across platforms from the _PowerShell_ side, it is certainly unexpected - and an artificial requirement - for anyone familiar with Unix shell scripting.

---

The bottom line is:

* **In an ideal world, PowerShell would transparently handle any _additional_, platform-specific escaping needs for interfacing with external programs _behind the scenes_, so that all that users need to take care of - on any supported platform - is to follow _PowerShell_'s escaping rules.**

* Sadly, this is not an option if backward compatibility must be preserved.

Version(s) of document impacted
------------------------------
- [x] Impacts 6.1 document
- [x] Impacts 6.0 document
- [x] Impacts 5.1 document
- [x] Impacts 5.0 document
- [x] Impacts 4.0 document
- [x] Impacts 3.0 document

**Comment by BrucePay** [comment]
@mklement0 If you invoke `powershell` from `cmd.exe` (or `bash` or whatever) the double-quoting issues come from the **calling language** not powershell. There is always a problem when you invoke one language from another because both languages use quoting. Try invoking `awk` from `bash` for example. The one scenario where powershell is involved as the calling language is powershell invoking powershell and in that scenario, we avoid the quoting nightmare by supporting scriptblock notation:
```
powershell.exe {
    Write-Host "This is a message"
}
```
(though I'm willing to bet that is not documented)

**Comment by mklement0** [comment]
@BrucePay 

> the double-quoting issues come from the calling language not powershell. 

That is _also_ true, but - on Windows - not _only_.

Yes, you first need to make the calling language happy so as not to cause a syntax error there.

But, due to the anarchy that is Windows argument passing, it is ultimately up to the _called_ program  to parse the command line as a _single string_.

As I've stated, that problem generally doesn't arise on Unix-like platforms, where programs are invoked with an _array_ of _literals_ - no command-line parsing needed (but see below).

PowerShell _is_ actively involved in transforming the command line, which is a _conceptual_ necessity on Windows, and an unfortunate _technical_ one on Unix-like platforms, due to current limitations in CoreFX that will be addressed in 2.1 (see below).

The way the transformation is implemented is unfortunate:

For instance:

```powershell
echoArgs.exe 'Nat "King" Cole'
```

results in the following command line:

```sh
"C:\path\to\EchoArgs.exe" "Nat "King" Cole"
```

While the `'...'` were translated into `"..."` - a sensible transformation, given that most utilities on Windows do not recognize `'...'` as string delimiters - _the embedded `"` were not escaped_, resulting in a broken command line.

To put it differently: 

PowerShell should not never have put the burden of `\`-escaping `"` chars. on the user, `echoArgs.exe 'Nat "King" Cole'` should be translated into `echoArgs.exe "Nat \"King\" Cole"` _behind the scenes_.  

While - as stated - it is ultimately up to the target program to parse the command line, enclosing arguments in `"..."` with embedded `"` chars. escaped as `\"` is the most widely used convention and - with the notable exception of batch files - works with external utilities of both Windows and Unix heritage.

Even PowerShell itself honors this convention when it is called from the outside: `\"` is treated as an escaped `"`, even though PowerShell-_internally_ `` ` ``, not `\` is the escape character.

On Unix-like platforms, PowerShell should only ever have passed the arguments resulting after _its_ parsing and expansion as an array of literals to the target program - the problem of escaping just for the _command line_ would never even arise.

Unfortunately, prior to the addition of [`ProcessStartInfo.ArgumentList` in 2.1](https://github.com/dotnet/corefx/issues/23592) it couldn't actually do that, but the way it manually constructs the command line has the same flaws as on Windows.

Yes, the behavior is consistent, but it is now broken on _all_ supported platforms.
And it sounds like that brokenness is here to stay, for fear of breaking existing scripts.

The pain of that will be felt more strongly on Unix-like platforms, where calling external utilities is much more common than on Windows.
(And the pain will also be felt when calling _into_ PowerShell, due to how `-Command` parsing works, but that's a separate story.)

It is one of the real pain points that motivated me to start the conversation about ["PowerShell vZeroTechnicalDebt"](https://github.com/PowerShell/PowerShell/issues/6745)

**In short: The need for the _user_ to `\`-escape `"` chars. when calling external programs is far from obvious and therefore needs documenting, which is why I opened this issue.**

---

Indeed, use of a script block  - available _from within PowerShell only_  and only _when calling PowerShell_ - is the best way to avoid quoting headaches in that scenario.

Its use is documented in `Get-Help about_PowerShell.exe` / `powershell -?`, but (a) not with respect to quoting and (b) it's a good idea to also mention it in the help topics mentioned above.

**Comment by bergmeister** [comment]
I should add that there are also scenarios where it is better to replace a double quote with 3 double quotes in the case of the [ProcessStartInfo.Arguments](https://msdn.microsoft.com/en-us/library/system.diagnostics.processstartinfo.arguments(v=vs.110).aspx) property in .Net, which boils down to be caused by the Windows API that is called at the end (therefore not really PowerShell's fault). The documentation on it says:
> A single argument that includes spaces must be surrounded by quotation marks, but those quotation marks are not carried through to the target application. In include quotation marks in the final parsed argument, triple-escape each mark.

**Comment by mklement0** [comment]
Thanks, @bergmeister, good to know, but that's a slightly different scenario in that the intent there is to pass _embedded, literal_ double quotes - in addition to satisfying _syntactic_ quoting requirements.

`ProcessStartInfo.Arguments` is itself _not_ plagued by the need for extra escaping; that is, if - taking the escaping needs of the calling language into account - you manage to pass a `"..."` string as part of a string's _content_, it _is_ passed through to the callee (or, in the case of Unix-like platforms, tokenized as expected; note that [.NET Core 2.1 will give you the option to alternatively pass arguments as an _array of literals_ via the new `.ArgumentList` property](https://github.com/dotnet/corefx/issues/23592).)

The triple-`"` technique surrounds an embedded `"..."` string - whose double quotes have _syntactic_ function from the perspective of the callee - with `""` on either end, with each of the latter representing an _embedded, literal_ `"` to the callee.

Unfortunately, however, the **triple-`"` technique doesn't work on _Unix_-like platforms.**

**What works on _both_ Windows and Unix is to use `\""...\""` rather than `"""..."""`**.
That is, the embedded literal quotes must be escaped as `\"`; e.g.:

```powershell
$psi = [System.Diagnostics.ProcessStartInfo]::new()
$psi.Arguments = 'one \""two in literal quotes"\"'
```

**Comment by sdwheeler** [comment]
@SteveL-MSFT is this worth trying to document?

**Comment by SteveL-MSFT** [comment]
@sdwheeler good question.  I think we are ok with just adding an example where we execute a native command with one that requires literal quotes.

**Comment by mklement0** [comment]
@SteveL-MSFT: 

Given how much of a pain point quoting for external programs is (including calling another instance of PowerShell), please consider a dedicated `about_*` topic that covers it; e.g. `about_Native_Quoting_Rules` (though I wonder if there's a better word than "native") - or at least add a dedicated section to `about_Quoting_Rules`.

As stated, with CLI-savvy Unix users starting to use PowerShell who expect friction-less calls to external utilities, the pain will increase - see https://github.com/PowerShell/PowerShell/issues/6935 for a recent example.

What's needed is a _systematic_ explanation of how quoting does _not_ work as it should with external programs, and how to work around that - not just an _example_ or two.

In addition to what I've discussed above, such a topic could cover additional pitfalls:

* That `Start-Process -ArgumentList` too requires an extra layer of embedded quoting - see https://github.com/PowerShell/PowerShell/issues/5576

* How compound tokens with quoted substrings are stringified; e.g.,
`msiexec INSTALLLOCATION='c:\tmp 1'` translates to `misexec "INSTALLLOCATION=c:\tmp 1"` behind the scenes (the argument _as a whole_ is enclosed in `"`), which breaks the command; - see [this SO post](https://github.com/PowerShell/PowerShell/issues/5576).

**Comment by SteveL-MSFT** [comment]
@mklement0 Would you be willing to submit a draft of this about topic?

**Comment by mklement0** [comment]
Thanks, @SteveL-MSFT, but I prefer that someone else take this on.

**Comment by 2BitSalute** [comment]
I would also like to see an example of passing in JSON (from a literal string or from a file) to an external program. It's a common enough pattern that it would be good to document for those looking for answers to why what they do in, e.g., bash, doesn't work in PowerShell.

**Comment by mklement0** [comment]
Good suggestion, @2BitSalute; in fact it [just came up on SO](https://stackoverflow.com/q/52822984/45375) (which may be how you found this post).

In short, you currently must **manually `\`-escape all embedded `"` instances in a JSON string** in order to pass it as an _argument_ to external program - which is both unexpected and a nuisance.
(Providing the string via stdin is not affected.)

```powershell
# Sample JSON
PS> $json = '{ "hello": "world" }'

# Pass to Bash for echoing the argument.
PS> bash -c 'echo $1' - $json
{ hello: world }   # !! Double quotes were stripped.

# Cumbersome workaround: \-escape embedded " chars.
PS> bash -c 'echo $1' - ($json -replace '"', '\"')
{ "hello": "world" }  # OK
```

On a side note, the Bash command should really be `'echo "$1"` for as-is echoing, but those `"` would get stripped too - with no _obvious_ breakage.

So, in summary, what _should_ work as follows:

```powershell
PS> bash -c 'echo "$1"' - $json
```

currently has to be:

```powershell
PS> bash -c 'echo \"$1\"' - ($json -replace '"', '\"')
```

Backward compatibility is the _only_ reason to hang on this behavior.

**Comment by aminya** [comment]
How can I write a command that works both on Windows and Linux? This is crucial for Github Actions workflow files.

**Comment by mklement0** [comment]
An update for those not following the discussion in https://github.com/PowerShell/PowerShell/issues/1995: 

https://github.com/PowerShell/PowerShell/issues/1995#issuecomment-673662322 states that no action will be taken to fix the underlying problem in v7.1

In the meantime, as an alternative to the currently required workarounds for passing arguments properly to external programs _from PowerShell_ - which would break if and when the underlying issues get fixed - you can use the **PSv3+ `ie` helper function** from the  **[`Native` module](https://github.com/mklement0/Native)** (in PSv5+, install with `Install-Module Native` from the [PowerShell Gallery](https://www.powershellgallery.com/packages/Native)), which internally compensates for all broken behavior and allows passing arguments as expected; to use it, simply prepend `ie` to your invocations; e.g.:  
`ie pwsh -noprofile -command ' "hi" '` works as expected.

**Comment by Luiz-Monad** [comment]
![escape](https://user-images.githubusercontent.com/4574729/122860339-27148f00-d2e3-11eb-9fcd-8862ae5614df.png)
sweet lord of escapes !
